### ISSUE TITLE
implement CNN for public opinion polarity text classification

### ISSUE THREAD

**Issue by jacksonwalters** [jacksonwalters/scotus-v-public#1]
Rather than have the metrics represent support in [0,1] it should measure polarity [-1,1]. We are measuring how strongly the opinion for or, crucially, against the issue is. Need to determine whether we're measuring conservative v. liberal points of view. The are other axes (pairs of opposites) one could consider, but this is the most straightforward, general, and relevant.

For public polarity, could take the following approach: the data we have is a question + a response, e.g.

**Q:** Should gays be allowed in the military?
**A:**

1. Strongly agree
2. Agree
3. Neutral
4. Disagree
5. Strongly disagree

(Q: "question")+ (A: "answer") = (S: "statement"). For instance, 

(Q: "Should gays be allowed in the military?") + (A: "Strongly agree") = (S: "I strongly agree that gays should be allowed in the military.").

We then perform a sentiment analysis on the statement and conclude that this a *liberal* statement, with a numeric value of, say, .8 on a scale of [-1,+1] with -1 being the most conservative and +1 being the most liberal.

**Comment by jacksonwalters** [comment]
This is pretty straightforward, and don't want to reinvent the wheel here as there are out of the box tools that will work fine for a prototype: [](https://github.com/jacksonwalters/cnn-text-political-bias-tf)

**Comment by jacksonwalters** [comment]
It appears the classifier from dennybritz has been broken (tf.contrib deprecated) by the upgrade to TensorFlow 2.0+. TensorFlow 1.14 is no longer available. They are trying to get people to move to pre-built datasets. It's easier to train a model, but harder to find the data you need. I trained a classifier quickly for sentiment analysis, but there seems to be no dataset for political bias. I could create a custom one, but that would take some time.

**Comment by jacksonwalters** [comment]
After playing around with a couple models, I decided to use a binary classifier with a 1d CNN layer using the Keras/tensorflow framework. It classifies sentences. Initially, I used a training set with 6,000 labeled tweets (each) from Trump=0 and Biden=1. The network does a reasonable job at picking out keywords and 2-3 word phrases which may appear in the tweets.

I hand labeled ~150 questions. In this case, the label represents whether an affirmative answer to the political opinion question indicates a liberal or conservative sentiment. The model can then be used to classify the remaining 900 or so questions. This resolves the issue.

**Note:** 150 questions is not enough to obtain accurate results. The classification is rather subtle (depends on long range correlations). Thousands of labeled questions are probably required. An RNN may also be better suited to this task.
